### PR TITLE
feature/relayserver: use eventbus.Monitor to simplify lifecycle management

### PR DESCRIPTION
**Pull request by creachadair**
Note to reviewers: This may be easier to read with whitespace disabled.

Instead of using separate channels to manage the lifecycle of the eventbus
client, use the recently-added eventbus.Monitor, which handles signaling the
processing loop to stop and waiting for it to complete.  This allows us to
simplify some of the setup and cleanup code in the relay server.

Updates #15160

Change-Id: Ia1a47ce2e5a31bc8f546dca4c56c3141a40d67af
Signed-off-by: M. J. Fromberger <fromberger@tailscale.com>

### CODE REVIEW

**Comment by creachadair** [comment]
@jwhited I tagged you here just for a double-check, I don't think this change has anything subtle but I wanted to make sure I've not missed something.

File: feature/relayserver/relayserver.go [hunk 1]
```diff
@@ -86,14 +85,13 @@ func (e *extension) handleBusLifetimeLocked() {
...
 		return // already running
 	}
-	port := *e.port
-	e.disconnectFromBusCh = make(chan struct{})
-	e.busDoneCh = make(chan struct{})
-	go e.consumeEventbusTopics(port)
+
+	e.eventSubs = new(eventbus.Monitor)
+	ec := e.bus.Client("relayserver.extension")
+	*e.eventSubs = ec.Monitor(e.consumeEventbusTopics(ec, *e.port))
```
Comment (nickkhyl) [inline feature/relayserver/relayserver.go]: Maybe use `ptr.To`?
```suggestion
	ec := e.bus.Client("relayserver.extension")
	e.eventSubs = ptr.To(ec.Monitor(e.consumeEventbusTopics(ec, *e.port)))
```

Comment (creachadair) [reply feature/relayserver/relayserver.go]: Good call, done.

File: feature/relayserver/relayserver_test.go [hunk 2]
```diff
@@ -157,13 +157,13 @@ func Test_extension_handleBusLifetimeLocked(t *testing.T) {
 				bus:                           eventbus.New(),
 				shutdown:                      tt.shutdown,
 				port:                          tt.port,
-				busDoneCh:                     tt.busDoneCh,
+				eventSubs:                     tt.eventSubs,
```
Comment (nickkhyl) [inline feature/relayserver/relayserver_test.go:160]: We don't necessarily need to do it in this PR, but based on usage I think the `eventSubs` test field is always `nil` and can be removed? Or did we mean to use it (or, formerly, `busDoneCh`) in any tests?

cc @jwhited

**Review by nickkhyl — APPROVED** [review]
LGTM, though `busDoneCh`/`eventSubs` usage in the old test looks a bit odd.

**Comment by creachadair** [comment]
> LGTM, though `busDoneCh`/`eventSubs` usage in the old test looks a bit odd.

Yeah, I noticed those tests don't actually exercise it at all, which is probably something we should improve on eventually.

File: feature/relayserver/relayserver_test.go (hunk 2 above)
Comment (creachadair) [reply feature/relayserver/relayserver_test.go:160]: I assumed it was added to allow for lifecycle to be tested, and that we just never got around to it. It's maybe worth adding at some point. For now I just left the pattern in place.

**Comment by jwhited** [comment]
https://github.com/tailscale/tailscale/pull/17239 adds a new channel to the select{} in `consumeEventbusTopics`, so I'm currently trying to wrap my head around `eventbus.Monitor` to see if these changes may conflict.

**Comment by creachadair** [comment]
> #17239 adds a new channel to the select{} in `consumeEventbusTopics`, so I'm currently trying to wrap my head around `eventbus.Monitor` to see if these changes may conflict.

I can wait on merging this and include your new subscription in a rebase, if you'd prefer.

**Comment by creachadair** [comment]
Marking this as draft for now, I will rebase after #17239 merges.

File: feature/relayserver/relayserver_test.go (hunk 2 above)
Comment (jwhited) [reply feature/relayserver/relayserver_test.go:160]: > I assumed it was added to allow for lifecycle to be tested, and that we just never got around to it.

This is most likely what happened. We can add a check in the test to make sure the pointer doesn't change when it's non-nil entering the func.

**Comment by creachadair** [comment]
@jwhited PTAL at your convenience

**Comment by creachadair** [comment]
Gentle reminder, @jwhited (this is still not urgent but I would like to get it settled at some point 🙂)
